File: rtc-over-websocket/js/JitterBuffer.js

class JitterBuffer {
    constructor(processLen,inputSamplerate,playSamperate,processBufSize,sonic) {
        this.playBuffer = new Array();

        this.stashBuffer = null;
        
        this.processLen = processLen;
        this.playBufLen = 2;
        this.inputSamplerate = inputSamplerate;
        this.playSamperate = playSamperate;
        this.sonic = sonic;
        this.delay = 0;
        this.processBufSize = processBufSize;

        this.stashLastFrame = null;

        this.isAccPlay = false;
        //this.test = false;
    }

    setDelay(delay){
        this.delay = delay;
    }

    appendBuffer(fromBuffer){
        let newBuffer = null;
        if(fromBuffer instanceof Float32Array){
            if(this.stashBuffer == null) this.stashBuffer = new Float32Array(0);
            newBuffer = new Float32Array(this.stashBuffer.length+fromBuffer.length);

        }else if(fromBuffer instanceof Int16Array){
            if(this.stashBuffer == null) this.stashBuffer = new Int16Array(0);
            newBuffer = new Int16Array(this.stashBuffer.length+fromBuffer.length);
        }
        
        newBuffer.set(this.stashBuffer, 0);
        newBuffer.set(fromBuffer, this.stashBuffer.length);
        this.stashBuffer = newBuffer;

        //console.log(this.stashBuffer.length);
    }

    extractBuffer(nSamples){
        let buf = null;
        if(this.stashBuffer!= null && this.stashBuffer.length >= nSamples){
            buf = this.stashBuffer.subarray(0, nSamples);
            this.stashBuffer = this.stashBuffer.subarray(nSamples);
        }
        return buf;
    }

    appendToPlayBuffer(){
        
        if(this.getLength() > (this.inputSamplerate*3/10) && this.isAccPlay == false){
            this.isAccPlay = true;
        }

        if(this.getLength() < (this.inputSamplerate/10) && this.isAccPlay == true){
            this.isAccPlay = false;
        }

        while(this.playBuffer.length < this.playBufLen){
            let blockBuffer = this.extractBuffer(this.processBufSize);
            if(blockBuffer != null){
                this.playBuffer.push(blockBuffer);
            }else{
                break;
            }
        }

    }

    push(buf){

        //console.log("push len:"+buf.length);
        this.appendBuffer(buf);
        this.appendToPlayBuffer();
        //console.log("this.stashBuffer:"+this.stashBuffer.length);
        
        
    }
    
    pop(){

        let buf = null;
        if(this.playBuffer.length > 0){
            buf = this.playBuffer.shift();
            this.stashLastFrame = buf;
        }else{
            this.appendToPlayBuffer();
            if(this.playBuffer.length > 0){
                buf = this.playBuffer.shift();
                this.stashLastFrame = buf;
            }
        } 
        this.appendToPlayBuffer();


        if(this.isAccPlay == true){
           console.log("Drop a audio frame");
           this.extractBuffer(this.processBufSize);
           if(this.getLength() < (this.inputSamplerate/10)){
            this.isAccPlay = false;
           }
       }
        
        if(buf == null && this.stashLastFrame != null){
            console.log("repeat audio frame!!")
            return this.stashLastFrame;
        }else{
            return buf;
        }
        
    }

    flushTempBuf(){

    }

    getLength(){
        return this.stashBuffer==null? 0:this.stashBuffer.length;
    }

    float2Int(floatArray){
        let intArray = new Int16Array(floatArray.length);
        for(let i =0; i<floatArray.length;i++){
            let s = floatArray[j];
            if (s < 0) {
                s = s * 32768;
            } else {
                s = s * 32767;
            }
            intArray[i] = Math.floor(s);
        }
        return intArray;
    }

    int2Float(intArray){
        let floatArray = new Float32Array(intArray.length);
        for (let i = 0; i < intArray.length; i++) {
            const sample = intArray[i];
            if(sample<0){
                floatArray[i] = sample / 32768; 
            }else{
                floatArray[i] = sample / 32767; 
            }
        }
        return floatArray;
    }

}

File: md_audio_websocket/android/app/src/main/kotlin/com/fsll/md_audio_websocket/JitterBuffer.kt
package org.fsll.jianghu3

class JitterBuffer(private val inputSamplerate: Int, private val processBufSize: Int){
    private var playBuffer: ArrayList<ByteArray> = ArrayList<ByteArray>(0)
    private var stashBuffer: ByteArray? = null
    private val playBufLen: Int = 2
    private var stashLastFrame: ByteArray? = null
    private var isAccPlay: Boolean = false

    private fun appendBuffer(fromBuffer: ByteArray) {
        var newBuffer: ByteArray?
        if (this.stashBuffer == null) this.stashBuffer = ByteArray(0)
        newBuffer = ByteArray(0)
        newBuffer = newBuffer.plus(this.stashBuffer!!)
        newBuffer = newBuffer.plus(fromBuffer)
        this.stashBuffer = newBuffer
    }

    private fun extractBuffer(nSamples: Int): ByteArray? {
        var buf: ByteArray? = null
        if (this.stashBuffer != null && this.stashBuffer!!.size >= nSamples) {
            buf = this.stashBuffer!!.copyOfRange(0, nSamples)
            this.stashBuffer = this.stashBuffer!!.copyOfRange(nSamples, this.stashBuffer!!.size)
        }
        return buf
    }

    private fun appendToPlayBuffer() {

        if (this.getLength() > (this.inputSamplerate * 3 / 10) && !this.isAccPlay) {
            this.isAccPlay = true
        }

        if (this.getLength() < (this.inputSamplerate / 10) && this.isAccPlay) {
            this.isAccPlay = false
        }

        while (this.playBuffer.size < this.playBufLen) {
            val blockBuffer = this.extractBuffer(this.processBufSize);
            if (blockBuffer != null) {
                this.playBuffer = this.playBuffer.plus(blockBuffer) as ArrayList<ByteArray>
            } else {
                break;
            }
        }

    }

    private fun getLength(): Int {
        return if (this.stashBuffer == null) 0 else this.stashBuffer!!.size
    }

    fun push(buf: ByteArray) {

        // console.log("push len:"+buf.length);
        this.appendBuffer(buf);
        this.appendToPlayBuffer();
        // console.log("this.stashBuffer:"+this.stashBuffer.length);


    }

    fun pop(): ByteArray? {
        var buf: ByteArray? = null
        if (this.playBuffer.isNotEmpty()) {
            buf = this.playBuffer[0]
            this.playBuffer.removeAt(0)
            this.stashLastFrame = buf
        } else {
            this.appendToPlayBuffer()
            if (this.playBuffer.size > 0) {
                buf = this.playBuffer[0]
                this.playBuffer.removeAt(0)
                this.stashLastFrame = buf
            }
        }
        this.appendToPlayBuffer();


        if (this.isAccPlay) {
            this.extractBuffer(this.processBufSize);
            if (this.getLength() < (this.inputSamplerate / 10)) {
                this.isAccPlay = false
            }
        }

        if (buf == null && this.stashLastFrame != null) {
            return null
        }
        return buf

    }

    fun clear() {
        playBuffer.clear()
        stashBuffer = null
        stashLastFrame = null
    }

}